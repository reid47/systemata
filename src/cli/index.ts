#!/usr/bin/env node
import * as path from 'path';
import * as fs from 'fs';
import { build } from './build';
import { version } from '../../package.json';
import { parseArgs, knownCommands } from './parse-args';
import { usage } from './usage';

const error = (message: string) => {
  console.error(`${message}\n\nUse '-h' or '--help' to show usage information.`);
  process.exit(1);
};

const { args, command } = parseArgs(process.argv.slice(2));

if (args.help) {
  console.log(usage);
  process.exit(0);
}

if (args.version) {
  console.log(version);
  process.exit(0);
}

if (command === 'init') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  if (fs.existsSync(path.resolve(process.cwd(), configFile as string))) {
    error(`File '${configFile}' already exists. Please specify a file name that doesn't exist.`);
  }
}

if (command === 'build') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  let config;
  try {
    config = require(path.resolve(process.cwd(), configFile as string));
  } catch (err) {
    error(`Error loading config file: ${err.message}`);
  }

  try {
    build(config);
    process.exit(0);
  } catch (err) {
    error(err);
  }
}

const validCommands = Object.keys(knownCommands).join(', ');

if (!command) {
  error(`No command specified. Command should be one of: ${validCommands}`);
}

error(`Invalid command specified: ${command}\nCommand should be one of: ${validCommands}`);
