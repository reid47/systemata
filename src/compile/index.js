import fs from 'fs';
import path from 'path';
import * as scss from './to-scss';

export default function compile(config) {
  const ext = scss.extension;
  fs.writeFileSync(
    path.resolve(__dirname, '../../ignored/wow' + ext),
    scss.compile(config),
    'utf8'
  );
}
