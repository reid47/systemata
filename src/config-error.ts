export class ConfigurationError extends Error {
  constructor(configPath: string, message: string) {
    const fullMessage = `Invalid configuration at "${configPath}": ${message}`;
    super(fullMessage);
  }
}
