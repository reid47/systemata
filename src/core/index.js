const { CssBuilder } = require('../css-builder');

function generate(config) {
  const { typography } = config;

  const cssBuilder = new CssBuilder();

  Object.keys(typography.sizes).forEach(typeSize => {
    cssBuilder.addRule(typeSize, {
      'font-size': typography.sizes[typeSize]
    });
  });

  console.log(JSON.stringify(config, null, 2));
  console.log(cssBuilder.toCss());
}

module.exports = { generate };
