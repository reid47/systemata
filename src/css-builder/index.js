class CssBuilder {
  constructor() {
    this.rules = {};
  }

  addRule(selector, styles) {
    this.rules[selector] = {
      ...this.rules[selector],
      ...styles
    };
  }

  toCss() {
    const lines = [];

    Object.keys(this.rules).forEach(selector => {
      lines.push(`${selector} {`);
      Object.keys(this.rules[selector]).forEach(property => {
        const value = this.rules[selector][property];
        lines.push(`  ${property}: ${value};`);
      });
      lines.push('}\n');
    });

    return lines.join('\n');
  }
}

module.exports = { CssBuilder };
