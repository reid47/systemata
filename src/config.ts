import * as path from 'path';
import {
  NamespaceConfig,
  Config,
  ColorConfig,
  PropertyMappingConfig,
  SpacingConfig,
  OutputConfig,
  SystemConfig,
  OutputFileConfig,
  DevelopmentConfig
} from './types';
import { version as systemataVersion } from '../package.json';
import { parseSemVer } from './utils';
import { ConfigurationError } from './config-error';

function resolveSystem(systemConfig: any): SystemConfig {
  systemConfig = systemConfig || {};

  const version = systemConfig.version || '0.0.1';
  if (!parseSemVer(version).valid) {
    throw new ConfigurationError(
      'system.version',
      `'${version}' is not a three-part version string (e.g. '0.1.2').`
    );
  }

  return {
    name: systemConfig.name || 'untitled-system',
    version,
    description: systemConfig.description || ''
  };
}

function resolveDevelopment(devConfig: any): DevelopmentConfig {
  devConfig = devConfig || {};

  return {
    serverPort: devConfig.serverPort || 4700
  };
}

function resolveOutput(outputConfig: any, configFileDir: string): OutputConfig {
  outputConfig = outputConfig || {};

  const files: OutputFileConfig[] = [];

  (outputConfig.files || []).forEach((file: any, index: number) => {
    file = file || {};

    const fileName = file.fileName || '';
    if (!fileName) {
      throw new ConfigurationError(`output.files[${index}].fileName`, 'File name must be specified.');
    }

    files.push({
      format: file.format || 'css',
      fileName
    });
  });

  return {
    buildDirectory: path.resolve(configFileDir, outputConfig.buildDirectory || './systemata-build'),
    archiveDirectory: path.resolve(configFileDir, outputConfig.archiveDirectory || './systemata-archive'),
    files
  };
}

function resolveNamespace(namespaceConfig: any): NamespaceConfig {
  namespaceConfig = namespaceConfig || {};

  return {
    prefix: namespaceConfig.prefix || false,
    parentClass: namespaceConfig.parentClass || false,
    peerClass: namespaceConfig.peerClass || false
  };
}

function resolvePropertyMapping(mappingsConfig: any): PropertyMappingConfig {
  mappingsConfig = (mappingsConfig || {}) as object;

  const defaultMappings: PropertyMappingConfig = {
    'background-color': 'bg',
    'border-color': 'brc',
    color: 'c',
    fill: 'fill',
    margin: 'm',
    'margin-bottom': 'mb',
    'margin-left': 'ml',
    'margin-right': 'mr',
    'margin-top': 'mt',
    'margin-top,margin-bottom': 'mv',
    'margin-left,margin-right': 'mh',
    padding: 'p',
    'padding-bottom': 'pb',
    'padding-left': 'pl',
    'padding-right': 'pr',
    'padding-top': 'pt',
    'padding-top,padding-bottom': 'pv',
    'padding-left,padding-right': 'ph'
  };

  return { ...defaultMappings, ...mappingsConfig };
}

function resolveColors(colorConfig: any): ColorConfig {
  if (!colorConfig) return {};
  return colorConfig as ColorConfig;
}

function resolveSpacing(spacingConfig: any): SpacingConfig {
  if (!spacingConfig) return {};
  return spacingConfig as SpacingConfig;
}

export function resolveConfig(config: any, configFilePath: string): Config {
  config = config || {};

  const configFileDir = path.dirname(configFilePath);

  return {
    __configFilePath: configFilePath,
    __systemataVersion: systemataVersion,
    system: resolveSystem(config.system),
    development: resolveDevelopment(config.development),
    namespace: resolveNamespace(config.namespace),
    propertyMapping: resolvePropertyMapping(config.propertyMapping),
    output: resolveOutput(config.output, configFileDir),
    color: resolveColors(config.color),
    spacing: resolveSpacing(config.spacing)
  };
}
