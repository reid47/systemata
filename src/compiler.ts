import { resolveConfig } from './config';
import { Config, OutputOptions } from './types';
import { generateCss } from './generate-css';

export class Compiler {
  config: Config;
  output: OutputOptions;

  constructor(config: any, output: OutputOptions) {
    this.config = resolveConfig(config);
    this.output = output;
  }

  toString() {
    if (this.output.format === 'css') return generateCss(this.config);

    throw new Error('Unsupported output configuration');
  }
}
