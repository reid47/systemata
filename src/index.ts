import { Compiler } from './compiler';

export function generate(config: any): string {
  return new Compiler(config).toString();
}
