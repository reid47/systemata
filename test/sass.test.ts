import { generate } from '../src/index';

const output = { format: 'sass' };

test('color styles', () => {
  const config = {
    output,
    color: {
      primary: '#f00',
      secondary: '#0f0'
    }
  };

  expect(generate(config)).toMatchSnapshot();
});
