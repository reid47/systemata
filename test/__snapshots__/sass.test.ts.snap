// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color styles 1`] = `
"$color-primary: #f00;
$color-secondary: #0f0;
.bg-primary { background-color: $color-primary }
.bg-secondary { background-color: $color-secondary }
.brc-primary { border-color: $color-primary }
.brc-secondary { border-color: $color-secondary }
.c-primary { color: $color-primary }
.c-secondary { color: $color-secondary }
.fill-primary { fill: $color-primary }
.fill-secondary { fill: $color-secondary }"
`;
